File: typescript/packages/core/src/Consolidation.ts
import { Effect } from "effect"
import { Record as AuraRecord, type FileWrite, type FileWriteError } from "@aura/contract"
import type { NGramIndex } from "@aura/indexing"
import * as Graph from "./Graph"

/**
 * Hard merge threshold — no LLM needed.
 *
 * @zh 硬合并阈值；不需要 LLM 参与。
 *
 * Rust reference: `CONSOLIDATION_THRESHOLD` (`../src/consolidation.rs`).
 */
export const CONSOLIDATION_THRESHOLD = 0.85

/**
 * Soft merge threshold — LLM-assisted range.
 *
 * @zh 软合并阈值；Rust 中为 LLM 辅助区间。
 *
 * Rust reference: `CONSOLIDATION_SOFT_THRESHOLD` (`../src/consolidation.rs`).
 */
export const CONSOLIDATION_SOFT_THRESHOLD = 0.5

/**
 * Result of a consolidation run.
 *
 * @zh 一次 consolidation 运行结果。
 *
 * Rust reference: `ConsolidationResult` (`../src/consolidation.rs`).
 */
export interface ConsolidationResult {
  readonly merged: number
  readonly checked: number
}

/**
 * Store operations needed by hard-merge consolidation.
 *
 * @zh 硬合并 consolidation 需要的认知存储操作。
 *
 * Rust reference: `CognitiveStore` usage in `graph::merge_records` (`../src/graph.rs`).
 */
export interface ConsolidationStore {
  readonly appendUpdate: (record: AuraRecord) => Effect.Effect<void, FileWriteError, FileWrite>
  readonly appendDelete: (id: string) => Effect.Effect<void, FileWriteError, FileWrite>
  readonly flush: () => Effect.Effect<void, FileWriteError, FileWrite>
}

/**
 * Build the Aura ID index used by consolidation.
 *
 * @zh 构建 consolidation 使用的 aura_id 到 record ID 索引。
 *
 * Rust reference: `Aura { aura_index }` and `graph::merge_records` (`../src/aura.rs`, `../src/graph.rs`).
 */
export function createAuraIndex(records: ReadonlyMap<string, AuraRecord>): Map<string, string> {
  const auraIndex = new Map<string, string>()
  for (const [recordId, record] of records) {
    if (typeof record.aura_id === "string" && record.aura_id.length > 0) {
      auraIndex.set(record.aura_id, recordId)
    }
  }
  return auraIndex
}

/**
 * Run hard-merge consolidation (MinHash >= 0.85).
 *
 * Finds duplicate pairs, keeps the higher-importance record, then merges
 * tags/connections/strength from the other record through `graph::merge_records`.
 *
 * @zh 查找重复 pair，保留 importance 更高的 record，并通过 `graph::merge_records`
 * 合并另一条 record 的 tags/connections/strength。
 *
 * Rust reference: `consolidation::consolidate` (`../src/consolidation.rs`).
 */
export function consolidate(
  records: Map<string, AuraRecord>,
  ngramIndex: NGramIndex,
  tagIndex: Map<string, Set<string>>,
  auraIndex: Map<string, string>,
  store: ConsolidationStore,
): Effect.Effect<ConsolidationResult, FileWriteError, FileWrite> {
  return Effect.gen(function* () {
    let merged = 0
    const nsMap = new Map<string, string>()
    for (const [id, record] of records) nsMap.set(id, record.namespace)

    const pairs = ngramIndex
      .findSimilarPairs(CONSOLIDATION_THRESHOLD)
      .filter(([idA, idB]) => nsMap.get(idA) === nsMap.get(idB))
    const checked = pairs.length
    const removed = new Set<string>()

    for (const [idA, idB] of pairs) {
      if (removed.has(idA) || removed.has(idB)) continue

      const recordA = records.get(idA)
      const recordB = records.get(idB)
      const impA = recordA === undefined ? 0 : AuraRecord.importance(recordA)
      const impB = recordB === undefined ? 0 : AuraRecord.importance(recordB)
      const [keepId, removeId] = impA >= impB ? [idA, idB] : [idB, idA]
      const removeBefore = records.get(removeId)

      const merge = yield* Graph.mergeRecords(keepId, removeId, records)
      replaceRecords(records, merge.records)

      if (removeBefore !== undefined) {
        removeFromIndexes(removeBefore, ngramIndex, tagIndex, auraIndex)
        yield* store.appendDelete(removeId)
        if (merge.keep !== null) {
          yield* store.appendUpdate(merge.keep)
        }
      }

      removed.add(removeId)
      merged += 1
    }

    if (merged > 0) {
      yield* store.flush()
    }

    return { merged, checked }
  })
}

function replaceRecords(target: Map<string, AuraRecord>, source: ReadonlyMap<string, AuraRecord>): void {
  target.clear()
  for (const [id, record] of source) target.set(id, record)
}

function removeFromIndexes(
  record: AuraRecord,
  ngramIndex: NGramIndex,
  tagIndex: Map<string, Set<string>>,
  auraIndex: Map<string, string>,
): void {
  ngramIndex.remove(record.id)
  for (const tag of record.tags) {
    tagIndex.get(tag)?.delete(record.id)
  }
  if (typeof record.aura_id === "string" && record.aura_id.length > 0) auraIndex.delete(record.aura_id)
}

File: src/consolidation.rs
//! MinHash deduplication & merge.
//!
//! Rewritten from aura-cognitive memory.py consolidate().

use crate::cognitive_store::CognitiveStore;
use crate::graph;
use crate::ngram::NGramIndex;
use crate::record::Record;
use std::collections::{HashMap, HashSet};

/// Hard merge threshold — no LLM needed.
pub const CONSOLIDATION_THRESHOLD: f32 = 0.85;
/// Soft merge threshold — LLM-assisted range.
pub const CONSOLIDATION_SOFT_THRESHOLD: f32 = 0.5;

/// Result of a consolidation run.
#[derive(Debug, Default)]
pub struct ConsolidationResult {
    pub merged: usize,
    pub checked: usize,
}

/// Run hard-merge consolidation (MinHash >= 0.85).
///
/// Finds duplicate pairs, keeps the higher-importance record,
/// merges tags/connections/strength from the other.
pub fn consolidate(
    records: &mut HashMap<String, Record>,
    ngram_index: &mut NGramIndex,
    tag_index: &mut HashMap<String, HashSet<String>>,
    aura_index: &mut HashMap<String, String>,
    store: &CognitiveStore,
) -> ConsolidationResult {
    let mut result = ConsolidationResult::default();

    // Build namespace lookup for O(1) filtering
    let ns_map: HashMap<&str, &str> = records
        .iter()
        .map(|(id, r)| (id.as_str(), r.namespace.as_str()))
        .collect();

    // Find similar pairs (global MinHash) and pre-filter to same-namespace only
    let all_pairs = ngram_index.find_similar_pairs(CONSOLIDATION_THRESHOLD);
    let pairs: Vec<_> = all_pairs
        .into_iter()
        .filter(|(id_a, id_b, _)| ns_map.get(id_a.as_str()) == ns_map.get(id_b.as_str()))
        .collect();
    result.checked = pairs.len();

    // Process pairs (avoid double-processing)
    let mut removed: HashSet<String> = HashSet::new();

    for (id_a, id_b, _sim) in &pairs {
        if removed.contains(id_a) || removed.contains(id_b) {
            continue;
        }

        let imp_a = records.get(id_a).map(|r| r.importance()).unwrap_or(0.0);
        let imp_b = records.get(id_b).map(|r| r.importance()).unwrap_or(0.0);

        let scar_a = records
            .get(id_a)
            .map(|r| r.route_state_class() == crate::record::RouteStateClass::Refuted)
            .unwrap_or(false);
        let scar_b = records
            .get(id_b)
            .map(|r| r.route_state_class() == crate::record::RouteStateClass::Refuted)
            .unwrap_or(false);

        // Scar protection: a Refuted consequence scar must never be the record
        // that gets removed in a merge — its record_id and consequence metadata
        // (situation/action/trust) must survive so consequence_verdict still sees
        // it. If BOTH sides are scars, leave them both untouched (merging would
        // destroy one scar's lived record).
        let (keep_id, remove_id) = if scar_a && scar_b {
            continue;
        } else if scar_a {
            (id_a.clone(), id_b.clone()) // keep the scar
        } else if scar_b {
            (id_b.clone(), id_a.clone()) // keep the scar
        } else if imp_a >= imp_b {
            (id_a.clone(), id_b.clone())
        } else {
            (id_b.clone(), id_a.clone())
        };

        graph::merge_records(
            &keep_id,
            &remove_id,
            records,
            ngram_index,
            tag_index,
            aura_index,
            store,
        );

        removed.insert(remove_id);
        result.merged += 1;
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_consolidation_threshold() {
        assert!(CONSOLIDATION_THRESHOLD > CONSOLIDATION_SOFT_THRESHOLD);
        assert!(CONSOLIDATION_THRESHOLD <= 1.0);
    }
}
